display/graph/tree/list.go

```diff
@@ -0,0 +1,190 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package tree
+
+import (
+	"fmt"
+	"strings"
+
+	d "github.com/apache/skywalking-cli/display/displayable"
+	"github.com/apache/skywalking-cli/graphql/schema"
+	"github.com/apache/skywalking-cli/graphql/trace"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/gizak/termui/v3/widgets"
+	"github.com/urfave/cli"
+
+	"github.com/apache/skywalking-cli/logger"
+)
+
+const DefaultPageSize = 15
+const keymap = " Keymap "
+const cc = "<C-c>"
+
+func DisplayList(ctx *cli.Context, displayable *d.Displayable) error {
+	data := displayable.Data.(schema.TraceBrief)
+	condition := displayable.Condition.(*schema.TraceQueryCondition)
+	if err := ui.Init(); err != nil {
+		logger.Log.Fatalf("failed to initialize termui: %v", err)
+	}
+	defer ui.Close()
+
+	list := widgets.NewList()
+	list.TitleStyle.Fg = ui.ColorRed
+	list.TextStyle = ui.NewStyle(ui.ColorYellow)
+	list.WrapText = false
+
+	tree := widgets.NewTree()
+	tree.TextStyle = ui.NewStyle(ui.ColorYellow)
+	tree.WrapText = false
+	tree.TitleStyle.Fg = ui.ColorRed
+
+	help := widgets.NewParagraph()
+	help.WrapText = false
+	help.Title = keymap
+	help.Text = `[k          ](fg:red,mod:bold) Scroll Up
+		[<Up>       ](fg:red,mod:bold) Scroll Up
+		[j          ](fg:red,mod:bold) Scroll Down
+		[<Down>     ](fg:red,mod:bold) Scroll Down
+		[<Ctr-b>    ](fg:red,mod:bold) list Page Up
+		[<Ctr-f>    ](fg:red,mod:bold) list Page Down           
+		[p          ](fg:red,mod:bold) list Page Up
+		[n          ](fg:red,mod:bold) list Page Down
+		[<Ctr-u>    ](fg:red,mod:bold) Scroll Half Page Up
+		[<Ctr-d>    ](fg:red,mod:bold) Scroll Half Page Down
+		[<Home>     ](fg:red,mod:bold) Scroll to Top
+		[<Enter>    ](fg:red,mod:bold) Show Trace
+		[<End>      ](fg:red,mod:bold) Scroll to Bottom
+		[q          ](fg:red,mod:bold) Quit
+		[<Ctr-c>    ](fg:red,mod:bold) Quit
+        `
+	draw(list, tree, help, data, 0, ctx, condition)
+	listenTracesKeyboard(list, tree, data, ctx, help, condition)
+
+	return nil
+}
+
+func draw(list *widgets.List, tree *widgets.Tree, help *widgets.Paragraph, data schema.TraceBrief, showIndex int,
+	ctx *cli.Context, condition *schema.TraceQueryCondition) {
+	x, y := ui.TerminalDimensions()
+
+	if data.Total != 0 {
+		var traceID = data.Traces[showIndex].TraceIds[0]
+		list.Title = fmt.Sprintf("[ %d/%d  %s]", *condition.Paging.PageNum, totalPages(data.Total), traceID)
+		nodes, serviceNames := getNodeData(ctx, traceID)
+		tree.Title = fmt.Sprintf("[%s]", strings.Join(serviceNames, "->"))
+		tree.SetNodes(nodes)
+		list.Rows = rows(data, x/4)
+	} else {
+		noData := "no data"
+		list.Title = noData
+		tree.Title = noData
+	}
+
+	list.SetRect(0, 0, x, y)
+	tree.SetRect(x/4, 0, x, y)
+	help.SetRect(x-x/7, 0, x, y)
+	tree.ExpandAll()
```

Comment:
The reason why we cannot switch the focus between the `list` and `tree` is that we manually set the rectangle area of them, by placing these widgets into containers, the focus can be switched between them, refer to `skywalking-cli/display/graph/linear/linear.go`

Reply:
ok. i will check.